stackslib/src/clarity_vm/database/mod.rs

```diff
@@ -24,22 +24,89 @@ use crate::chainstate::burn::db::sortdb::{
     get_ancestor_sort_id, get_ancestor_sort_id_tx, SortitionDB, SortitionDBConn, SortitionHandle,
     SortitionHandleConn, SortitionHandleTx,
 };
-use crate::chainstate::nakamoto::NakamotoChainState;
+use crate::chainstate::nakamoto::{nakamoto_keys, NakamotoChainState, StacksHandle};
 use crate::chainstate::stacks::boot::PoxStartCycleInfo;
 use crate::chainstate::stacks::db::accounts::MinerReward;
 use crate::chainstate::stacks::db::{
-    ChainstateTx, MinerPaymentSchedule, StacksChainState, StacksHeaderInfo,
+    ChainstateTx, MinerPaymentSchedule, StacksChainState, StacksDBConn, StacksDBTx,
+    StacksHeaderInfo,
 };
 use crate::chainstate::stacks::index::marf::{MarfConnection, MARF};
 use crate::chainstate::stacks::index::{ClarityMarfTrieId, MarfTrieId, TrieMerkleProof};
 use crate::chainstate::stacks::Error as ChainstateError;
 use crate::clarity_vm::special::handle_contract_call_special_cases;
 use crate::core::{StacksEpoch, StacksEpochId};
-use crate::util_lib::db::{DBConn, FromColumn, FromRow};
+use crate::util_lib::db::{DBConn, Error as DBError, FromColumn, FromRow};
 
 pub mod marf;
 
-pub struct HeadersDBConn<'a>(pub &'a Connection);
+pub trait GetTenureStartId {
+    fn get_tenure_block_id(
+        &self,
+        tip: &StacksBlockId,
+        tenure_id_consensus_hash: &ConsensusHash,
+    ) -> Result<Option<TenureBlockId>, DBError>;
+    fn conn(&self) -> &Connection;
+}
+
+impl GetTenureStartId for StacksDBConn<'_> {
```

Comment:
Can you just implement once for the trait?
```rust
impl <SH: StacksHandle> GetTenureStartId for SH {
}
```

Reply:
Sure

Reply:
Actually, we can't -- the mutability of `self` differs